### PR TITLE
Cache more queries on disk

### PR DESCRIPTION
**Pull request by cjgillot**
One of the principles of incremental compilation is to allow saving results on disk to avoid recomputing them.
This PR investigates persisting a lot of queries whose result are to be saved into metadata.
Some of the queries are cheap reads from HIR, but we may also want to get rid of these reads for incremental lowering.

### CODE REVIEW

File: compiler/rustc_middle/src/query/mod.rs
Comment (bjorn3) [inline compiler/rustc_middle/src/query/mod.rs:1370]: Aren't those just a cheap hir lookup?

Comment (cjgillot) [reply compiler/rustc_middle/src/query/mod.rs:1370]: Yes. For the moment, I'm trying some heuristic like "cache everything that will be stored in metadata (roughly if is has separate_provide_extern)". I plan to refine this from perf measurements.

**Comment by cjgillot** [comment]
r? rust-lang/compiler

**Review by davidtwco — APPROVED** [review]
r=me if you are looking to land this and it isn't just an experiment

**Comment by cjgillot** [comment]
Note: the on-disk query result cache is significantly larger with this PR. Depending on the benchmark, the increase ranges 20-40%.

@davidtwco I am indeed looking to land this. I wonder if we should refine the disk space-run time tradeoff beforehand.

**Comment by davidtwco** [comment]
> @davidtwco I am indeed looking to land this. I wonder if we should refine the disk space-run time tradeoff beforehand.

Yeah, I think that makes sense. Perhaps nomination for a meeting makes sense.

**Comment by apiraino** [comment]
The cache disk space topic was discussed at length during the T-compiler meeting, though probably a follow-up meeting will be needed ([Zulip discussion](https://zulip-archive.rust-lang.org/stream/238009-t-compiler/meetings/topic/.5Bweekly.5D.202022-04-07.html#278173925)).

I'll remove the nomination (if it's ok)

@rustbot label -I-compiler-nominated

**Comment by JohnCSimon** [comment]
Ping from triage: 
@cjgillot merge conflicts

**Comment by cjgillot** [comment]
After investigation, I'm not sure how I can optimize the disk effect.  One of the key effects is to avoid recomputing `resolve_lifetimes` (for instance [here](https://perf.rust-lang.org/detailed-query.html?sort_idx=11&commit=36ace5e9edc63601569a7ef878ded180015f85d3&base_commit=e2301ca54320659835467072f37201591db959b5&benchmark=serde-1.0.136-check&scenario=incr-unchanged)).  This is done by skipping all "astconv" queries which access it indirectly. I'd rather avoid caching `resolve_lifetimes`, preferring queries whose results will be written into metadata.

**Review by davidtwco — APPROVED** [review]

**Comment by davidtwco** [comment]
If this is waiting on another review from me then all the changes in this PR still look good to me, so r=me if you don't want to do further changes.

> After investigation, I'm not sure how I can optimize the disk effect. One of the key effects is to avoid recomputing `resolve_lifetimes` (for instance [here](https://perf.rust-lang.org/detailed-query.html?sort_idx=11&commit=36ace5e9edc63601569a7ef878ded180015f85d3&base_commit=e2301ca54320659835467072f37201591db959b5&benchmark=serde-1.0.136-check&scenario=incr-unchanged)). This is done by skipping all "astconv" queries which access it indirectly. I'd rather avoid caching `resolve_lifetimes`, preferring queries whose results will be written into metadata.

I'm not familiar enough with the relevant code to have any suggestions here unfortunately.

**Comment by cjgillot** [comment]
Let's confirm there are no unforeseen regressions, and I think we'll go ahead.

@bors try @rust-timer queue

**Comment by cjgillot** [comment]
This PR caches many more queries' results on disk.  The queries have been chosen to be: (1) outputs of typechecking, and (2) parts of metadata.  This caching allows for wide and large compile time savings, up to 36%.  The tradeoff is a compile time increase (4%) when these caches cannot be usefully used, and a increased on-disk size around 30%.
@rustbot label: +pref-regression-triaged

**Pull request merged**
